### PR TITLE
add flag to flattenSpec to keep null columns

### LINKED ISSUE #9658: Sampler with Inline input source swallows up columns that are all `null`
### Affected Version

0.17.0

### Description

When posting to the sampler with an inline input source that has a column that is all `null` that column would not be returned as part of the parsed fields without explicit dimensions being set.

Try sending:

```json
{
  "type": "index",
  "spec": {
    "ioConfig": {
      "type": "index",
      "inputSource": {
        "type": "inline",
        "data": "{\"make\":\"Honda\",\"model\":\"Odyssey\",\"a\":null}\n{\"make\":\"Tesla\",\"model\":\"Cybertruck\",\"a\":null}"
      },
      "inputFormat": {
        "type": "json"
      }
    },
    "dataSchema": {
      "dataSource": "sample",
      "timestampSpec": {
        "column": "!!!_no_such_column_!!!",
        "missingValue": "1970-01-01T00:00:00Z"
      },
      "dimensionsSpec": {}
    },
    "type": "index",
    "tuningConfig": {
      "type": "index"
    }
  },
  "samplerConfig": {
    "numRows": 500,
    "timeoutMs": 15000
  }
}
```

to:

`/druid/indexer/v1/sampler`

You get back:

```json
{
  "numRowsRead": 2,
  "numRowsIndexed": 2,
  "data": [
    {
      "input": {
        "make": "Honda",
        "model": "Odyssey",
        "a": null
      },
      "parsed": {
        "__time": 0,
        "model": "Odyssey",
        "make": "Honda"
      }
    },
    {
      "input": {
        "make": "Tesla",
        "model": "Cybertruck",
        "a": null
      },
      "parsed": {
        "__time": 0,
        "model": "Cybertruck",
        "make": "Tesla"
      }
    }
  ]
}
```

Notice how the `a` column is in not returned as part of a the parsed section.

This creates issues with columns that happen to be all `null` in the sample (of 20) that is used by the console.

Just to be clear: I would expect the column to show up in the `parsed` section even if all the values for it are `null`

### PR DESCRIPTION
**Pull request by mcbrewster**
Fixes #9658.
Screen Shot 2020-05-03 at 2 52 34 PM

### Description

When The JsonReader flattens input it filters out key value pairs with null values, which prevents any all null column from being included as columns if no dimension spec is provided. This is a problem in the sampler because often all the entries in the sample will be null, meaning columns will be lost after flattening and wont appear in the web-console. The goal of this pr is to prevent the JSONFlattenMaker from removing null values from the document by adding a boolean flag keepNullColumns to the flattenSpec. This can then be added to the sampleSpec in the web console so that all null columns still appear in the ingestion flow.

I added the Json property keepNullColumns to JsonPathSpec (this is the flattenSpec), as well as the method isKeepNullColumns() which returns the values.  KeepNullColumns defaults to false if not provided, meaning there should be no change to the current function unless the user explicitly sets it to true. 

The JSONFlattenMaker class was modified to have a constructor that accepts keepNullColumns as a boolean value and is use to prevent discoverRootFields from filtering out null values. 

JSONReader passes keepNullColumns to the constructor of JSONFlattenMaker using the isKeepNullColumns() method. 

A test was added to JSONReaderTest to check that all columns including columns with null values are present as dimensions as well as flattening still works. 

Literally every other file is just adding the value false to JSONPathSpecs in tests

This PR has:
- [x ] been self-reviewed.
- [x ] added unit tests or modified existing tests to cover new code paths.
- [x] been tested in a test Druid cluster.

##### Key changed/added classes in this PR
 * `JsonReader`
 * `JSONFlattenMaker`
 * `JSONPathSpec`

### CODE REVIEW

File: core/src/main/java/org/apache/druid/java/util/common/parsers/JSONPathSpec.java [hunk 1]
```diff
@@ -28,19 +28,23 @@
...
   private final boolean useFieldDiscovery;
   private final List<JSONPathFieldSpec> fields;
+  private final boolean keepNullColumns;
 
   @JsonCreator
   public JSONPathSpec(
       @JsonProperty("useFieldDiscovery") Boolean useFieldDiscovery,
-      @JsonProperty("fields") List<JSONPathFieldSpec> fields
+      @JsonProperty("fields") List<JSONPathFieldSpec> fields,
+      @JsonProperty("keepNullColumns") Boolean keepNullColumns
```
Comment (jihoonson) [inline core/src/main/java/org/apache/druid/java/util/common/parsers/JSONPathSpec.java:41]: This property is used only in `JSONFlattenMaker` while this class is used by all nested formats. It would be better to have it in `JsonInputFormat` and `JSONParseSpec` instead.

File: core/src/main/java/org/apache/druid/java/util/common/parsers/JSONFlattenerMaker.java [hunk 2]
```diff
@@ -53,16 +53,23 @@
...
+
   @Override
   public Iterable<String> discoverRootFields(final JsonNode obj)
   {
     return FluentIterable.from(() -> obj.fields())
                          .filter(
                              entry -> {
                                final JsonNode val = entry.getValue();
-                               return !(val.isObject() || val.isNull() || (val.isArray() && !isFlatList(val)));
+                               return !(val.isObject() || (!keepNullValues && val.isNull()) || (val.isArray() && !isFlatList(val)));
```
Comment (jihoonson) [inline core/src/main/java/org/apache/druid/java/util/common/parsers/JSONFlattenerMaker.java:72]: I'm wondering why `JSONFlattenMaker` filters nulls out even though other `FlattenMaker` implementations don't. We should probably make the behavior consistent, but it should be done in a separate PR.

File: core/src/test/java/org/apache/druid/data/input/impl/JsonReaderTest.java [hunk 3]
```diff
@@ -132,4 +135,45 @@ public void testParseRowWithConditional() throws IOException
...
+        source,
+        null
+    );
+    final int numExpectedIterations = 1;
+    try (CloseableIterator<InputRow> iterator = reader.read()) {
+      int numActualIterations = 0;
+      while (iterator.hasNext()) {
+        final InputRow row = iterator.next();
+        Assert.assertEquals(Arrays.asList("path_omg", "timestamp", "bar", "foo"), row.getDimensions());
+        Assert.assertEquals(Collections.emptyList(), row.getDimension("bar"));
```
Comment (jihoonson) [inline core/src/test/java/org/apache/druid/data/input/impl/JsonReaderTest.java:172]: nit: can be `Assert.assertTrue(row.getDimension("bar").isEmpty());`

File: core/src/test/java/org/apache/druid/data/input/impl/JsonReaderTest.java [hunk 4]
```diff
@@ -132,4 +135,45 @@ public void testParseRowWithConditional() throws IOException
...
+        Assert.assertEquals("x", Iterables.getOnlyElement(row.getDimension("foo")));
+        Assert.assertEquals(Collections.emptyList(), row.getDimension("path_omg"));
+        numActualIterations++;
```
Comment (jihoonson) [inline core/src/test/java/org/apache/druid/data/input/impl/JsonReaderTest.java:175]: You probably want to verify how many rows it has read?

File: core/src/main/java/org/apache/druid/java/util/common/parsers/JSONFlattenerMaker.java (hunk 2 above)
Comment (jihoonson) [reply core/src/main/java/org/apache/druid/java/util/common/parsers/JSONFlattenerMaker.java:72]: This comment is not a blocker for this PR, but more like a note.

File: core/src/main/java/org/apache/druid/data/input/impl/JSONParseSpec.java [hunk 5]
```diff
@@ -37,46 +37,48 @@
 {
   private final ObjectMapper objectMapper;
   private final Map<String, Boolean> featureSpec;
+  private final Boolean keepNullColumns;
```
Comment (suneet-s) [inline core/src/main/java/org/apache/druid/data/input/impl/JSONParseSpec.java:40]:
```suggestion
  private final boolean keepNullColumns;
```

This way the class does not need to worry about nullability issues everywhere, it can just be handled in the constructor.

File: core/src/main/java/org/apache/druid/data/input/impl/JSONParseSpec.java [hunk 6]
```diff
@@ -98,7 +106,7 @@ public boolean equals(final Object o)
       return false;
     }
     final JSONParseSpec that = (JSONParseSpec) o;
-    return Objects.equals(featureSpec, that.featureSpec);
+    return Objects.equals(featureSpec, that.featureSpec) && Objects.equals(keepNullColumns, that.keepNullColumns);
```
Comment (suneet-s) [inline core/src/main/java/org/apache/druid/data/input/impl/JSONParseSpec.java:109]: Missing corresponding update to hashCode. Can you add an `EqualsVerifier` test for this class please.

File: core/src/main/java/org/apache/druid/data/input/impl/JSONParseSpec.java [hunk 7]
```diff
@@ -85,6 +87,12 @@ public ParseSpec withDimensionsSpec(DimensionsSpec spec)
     return featureSpec;
   }
 
+  @JsonProperty
+  public Boolean getKeepNullColumns()
```
Comment (jihoonson) [inline core/src/main/java/org/apache/druid/data/input/impl/JSONParseSpec.java:91]: Can be boolean here too.

File: web-console/src/utils/sampler.ts [hunk 8]
```diff
@@ -198,6 +201,8 @@ function makeSamplerIoConfig(
   } else if (specType === 'kinesis') {
     ioConfig = deepSet(ioConfig, 'useEarliestSequenceNumber', sampleStrategy === 'start');
   }
+  // In order to prevent potential data loss null columns should be kept by the sampler and shown in the ingestion flow
+  ioConfig = deepSet(ioConfig, 'inputFormat.keepNullColumns', true);
```
Comment (suneet-s) [inline web-console/src/utils/sampler.ts:205]: Does this mean it's hard coded to true on the web console? Is there ever a case we want it to be false?

File: core/src/test/java/org/apache/druid/data/input/impl/InputRowParserSerdeTest.java [hunk 9]
```diff
@@ -52,7 +52,8 @@ public void testStringInputRowParserSerde() throws Exception
             new TimestampSpec("timestamp", "iso", null),
             new DimensionsSpec(DimensionsSpec.getDefaultSchemas(ImmutableList.of("foo", "bar")), null, null),
             null,
-            null
+            null,
+            false
```
Comment (suneet-s) [inline core/src/test/java/org/apache/druid/data/input/impl/InputRowParserSerdeTest.java:56]: nit: it would be easier to read if you made false a constant at the top of this test, like `DO_NOT_KEEP_NULL_COLUMNS`

File: core/src/test/java/org/apache/druid/data/input/impl/JsonReaderTest.java [hunk 10]
```diff
@@ -132,4 +135,46 @@ public void testParseRowWithConditional() throws IOException
...
+            ImmutableList.of(
+                new JSONPathFieldSpec(JSONPathFieldType.PATH, "path_omg", "$.o.mg")
+            )
+        ),
+        null,
+        true
+    );
+
+    final ByteEntity source = new ByteEntity(
+        StringUtils.toUtf8("{\"timestamp\":\"2019-01-01\",\"bar\":null,\"foo\":\"x\",\"o\":{\"mg\":null}}")
```
Comment (suneet-s) [inline core/src/test/java/org/apache/druid/data/input/impl/JsonReaderTest.java:154]: Is there a test where keepNull is true, but there aren't any null columns?

File: server/src/test/java/org/apache/druid/segment/realtime/plumber/RealtimePlumberSchoolTest.java [hunk 11]
```diff
@@ -161,6 +162,7 @@ public void setUp() throws Exception
                     new TimestampSpec("timestamp", "auto", null),
                     new DimensionsSpec(null, null, null),
                     null,
+                    null,
```
Comment (suneet-s) [inline server/src/test/java/org/apache/druid/segment/realtime/plumber/RealtimePlumberSchoolTest.java:165]: WOW - I didn't realize we used so many constructors! Maybe one day, I'll get around to using mockito to make this refactoring less painful in the future.

File: web-console/src/utils/sampler.ts (hunk 8 above)
Comment (mcbrewster) [reply web-console/src/utils/sampler.ts:205]: Yes and no. It is hard coded to true on the sampleSpec but not the spec being created by the web console that is ingested into druid. This changes the web console so that all the columns including those that are null for all the values in the sample are shown to the user and keep during the ingestion flow, and will end up on the task spec that is generated by the web console. This prevents columns that only have null values from being excluded from being removed by the web console if it has data that just doesn't appear in the sample. So we would never want keepNullColumns be false on the sampleSpec because it could cause data to be lost. (you can actually see this happening in the changes I made to the e2e test for the web console)

File: core/src/main/java/org/apache/druid/java/util/common/parsers/JSONFlattenerMaker.java [hunk 12]
```diff
@@ -53,16 +53,24 @@
...
+
   @Override
   public Iterable<String> discoverRootFields(final JsonNode obj)
   {
     return FluentIterable.from(() -> obj.fields())
                          .filter(
                              entry -> {
                                final JsonNode val = entry.getValue();
-                               return !(val.isObject() || val.isNull() || (val.isArray() && !isFlatList(val)));
+                               // If the keepNullValues is set on the JSONParseSpec then null values should not be filtered out
```
Comment (suneet-s) [inline core/src/main/java/org/apache/druid/java/util/common/parsers/JSONFlattenerMaker.java:72]:
```suggestion
                               // Root fields are any json elements that ... 
```

I meant something like spelling out how we discover root fields. There are a bunch of nots and ands and ors that I can't come up with a statement that clearly defines what this is trying to do.

This really isn't part of your change, but if you figured out what this is actually doing, you'll help all future devs with a comment

**Comment by vogievetsky** [comment]
I looked over the TS changes and those look good to me 👍
